File: dotnet/mpmt1.cs
/*
 * mpmt1.cs: A stupid simple example of C# threading
 *
 * License:
 *   Apache License, Version 2.0
 * History:
 *   2024/05/26 v0.1 Initial version based on:
 * Author:
 *   Masanori Itoh <masanori.itoh@gmail.com>
 * Usage:
 *   * prepare dotnet sdk/runtime
 *      * e.g., mcr.microsoft.com/dotnet/sdk:8.0
 *   * `$ cd dotnet`
 *   * `$ dotnet restore`
 *   * `$ dotnet build`
 *   * `$ bin/Debug/net8.0/mpmt1 NUM_CONTEXT DURATION`
 * TODO:
 *   * Use getopt
 * References:
 *   * https://www.geeksforgeeks.org/c-sharp-multithreading/
 */
using System;
using System.Threading;
using System.Collections.Generic;

class Program
{
  static void busy_worker(object duration) {
    Console.WriteLine("busy_worker() called."); 
    long start_ts = DateTimeOffset.Now.ToUnixTimeMilliseconds();
    while (true) {
      long current_ts = DateTimeOffset.Now.ToUnixTimeMilliseconds();
      if (current_ts - start_ts > (long)duration) {
        Console.WriteLine("Expired.");
        break; 
      }
    }
    return;
  }

  static void Main(string[] args)
  {
    List<Thread>tlist = new List<Thread>();
    long duration = 5;
    long num_context = 3;

    if (args.Length >= 1) {
      num_context = Int32.Parse(args[0]);
    }
    if (args.Length >= 2) {
      duration = Int32.Parse(args[1]);
    }
    Console.WriteLine("NUM_CONTEXT: {0} DURATION: {1}", num_context, duration);

    for (int i = 0; i < num_context; i++) {
      tlist.Add(new Thread(busy_worker)); 
    }

    foreach (Thread t in tlist) {
      t.Start(duration * 1000);
    }

    foreach (Thread t in tlist) {
      t.Join();
    }   
  }
}

File: dart/mpmt1.dart
//
//  mpmt1.dart: A stupid simple example of Dart
//
// License:
//   Apache License, Version 2.0
// History:
//   * 2024/05/24 v0.1 Initial version
// Author:
//   Masanori Itoh <masanori.itoh@gmail.com>
// Usage:
//-  Prepare Dart sdk environemt such as docker Dart image
//-  $ dart pub get
//-  $ dart mpmt1.dart NUM_CONTEXT (number of threads) DURATION (in sec.)
// TODO:
//   * Use args
//   * Implement worker and main threads synchroniation
//
import 'package:sprintf/sprintf.dart';
import 'dart:async';
import 'dart:isolate';

void busy_worker(int duration_ms) async {    
    print("busy_worker() called.");
    int start_ts = DateTime.now().millisecondsSinceEpoch;
    while (true) {
        int current_ts = DateTime.now().millisecondsSinceEpoch;
        if ((current_ts - start_ts) >= duration_ms) {
            print("Expired.");
            break;
        }
    }
    return;
}

void main(List<String> args) async {
    int num_context = 3;
    int duration = 5;

    if (args.length >= 1) {
        num_context = int.parse(args[0]);
    }
    if (args.length >= 2) {
        duration = int.parse(args[1]);
    }
    
    print(sprintf("NUM_CONTEXT: %d DURATION: %d", [num_context, duration]));

    List<Future<int>> futures = [];
    for (int i = 0; i < num_context; i++) {
        Isolate.spawn(busy_worker, duration * 1000);
    }
    new Future.delayed(new Duration(seconds: duration + 1 )).then((_) => print("completed."));
}
    